examples/async-multisig/src/lib.rs

```diff
@@ -0,0 +1,129 @@
+#![no_std]
+
+use codec::{Decode, Encode};
+use core::str;
+use futures::{future, FutureExt};
+use gstd::{debug, msg, prelude::*, ProgramId};
+use gstd_async::msg as msg_async;
+use scale_info::TypeInfo;
+use sp_core::{
+    crypto::UncheckedFrom,
+    sr25519::{Pair as Sr25519Pair, Public, Signature},
+    Pair,
+};
+
+static mut SIGNATORIES: Vec<ProgramId> = vec![];
+static mut SIGNED_MESSAGE_PROGRAM: ProgramId = ProgramId([0u8; 32]);
+static mut THRESHOLD: usize = 0;
+
+const GAS_LIMIT: u64 = 1_000_000_000;
+
+#[derive(Debug, Encode, TypeInfo)]
+struct SignRequest {
+    message: Vec<u8>,
+}
+
+#[derive(Debug, Decode, TypeInfo)]
+pub struct SignResponse {
+    pub message: Vec<u8>,
+    pub signature: Vec<u8>,
+}
+
+gstd::metadata! {
+    title: "demo async multisig",
+    init:
+        input: Vec<u8>,
+    handle:
+        input: Vec<u8>,
+}
+
+fn hex_to_id(hex: &str) -> Result<ProgramId, ()> {
+    let hex = hex.strip_prefix("0x").unwrap_or(hex);
+
+    hex::decode(hex)
+        .map(|bytes| ProgramId::from_slice(&bytes))
+        .map_err(|_| ())
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn init() {
+    let input = String::from_utf8(msg::load_bytes()).expect("Invalid message: should be utf-8");
+    let dests: Vec<&str> = input.split(',').collect();
+    let len = dests.len();
+    if len < 3 {
+        panic!("Invalid input, should be a number and at least two IDs separated by comma");
+    }
+
+    THRESHOLD = usize::from_str_radix(dests[0], 10)
+        .map(|t| if t > len - 2 { len - 2 } else { t })
+        .map(|t| if t < 1 { 1 } else { t })
+        .expect("INTIALIZATION FAILED: INVALID THRESHOLD");
+
+    SIGNED_MESSAGE_PROGRAM = hex_to_id(dests[1]).expect("INTIALIZATION FAILED: INVALID PROGRAM ID");
+
+    SIGNATORIES = dests
+        .into_iter()
+        .skip(2)
+        .map(|s| hex_to_id(s).expect("INTIALIZATION FAILED: INVALID ACCOUNT ID"))
+        .collect();
+}
+
+#[gstd_async::main]
+async fn main() {
+    let message = msg::load_bytes();
+    debug!("message = {:?}", message);
+
+    let encoded = SignRequest {
+        message: message.clone(),
+    }
+    .encode();
+
+    let mut requests: Vec<_> = unsafe { &SIGNATORIES }
+        .iter()
+        .enumerate()
+        .map(|(i, s)| {
+            msg_async::send_and_wait_for_reply(*s, &encoded, GAS_LIMIT, 0).map(move |r| (i, r))
+        })
+        .collect();
+
+    let mut threshold = 0usize;
+    while !requests.is_empty() {
```

Comment:
This will send messages one-by-one

We need to send them all at once

Reply:
The following setlect_all line seems to send messages simultaneously

Reply:
oh nice trick, I couldn't even read it at first

seems ok, but will require quite a good amount of tests to show it's working in many situations

Reply:
sure, working on it